registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp

```diff
@@ -0,0 +1,174 @@
+                             const pcl::Correspondences &correspondences,
+                             Matrix4 &transformation_matrix) const
+{
+  ConstCloudIterator<PointSource> source_it (cloud_src, correspondences, true);
+  ConstCloudIterator<PointTarget> target_it (cloud_tgt, correspondences, false);
+  estimateRigidTransformation (source_it, target_it, transformation_matrix);
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointSource, typename PointTarget, typename Scalar> inline void
+pcl::registration::TransformationEstimationSymmetricPointToPlaneLLS<PointSource, PointTarget, Scalar>::
+constructTransformationMatrix (const Vector6 &parameters,
+                               Matrix4 &transformation_matrix) const
+{
+  // Construct the transformation matrix from rotation and translation 
+  const Eigen::AngleAxis<Scalar> rotation_z (parameters (2), Eigen::Matrix<Scalar, 3, 1>::UnitZ ());
+  const Eigen::AngleAxis<Scalar> rotation_y (parameters (1), Eigen::Matrix<Scalar, 3, 1>::UnitY ());
+  const Eigen::AngleAxis<Scalar> rotation_x (parameters (0), Eigen::Matrix<Scalar, 3, 1>::UnitX ());
+  const Eigen::Translation<Scalar, 3> translation (parameters (3), parameters (4), parameters (5));
+  const Eigen::Transform<Scalar, 3, Eigen::Affine> transform = rotation_z * rotation_y * rotation_x * translation * rotation_z * rotation_y * rotation_x;
+  transformation_matrix = transform.matrix();
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointSource, typename PointTarget, typename Scalar> inline void
+pcl::registration::TransformationEstimationSymmetricPointToPlaneLLS<PointSource, PointTarget, Scalar>::
+estimateRigidTransformation (ConstCloudIterator<PointSource>& source_it, ConstCloudIterator<PointTarget>& target_it, Matrix4 &transformation_matrix) const
+{
+  using Matrix6 = Eigen::Matrix<Scalar, 6, 6>;
+  using Vector3 = Eigen::Matrix<Scalar, 3, 1>;
+
+  Matrix6 ATA;
+  Vector6 ATb;
+  ATA.setZero ();
+  ATb.setZero ();
+  auto M = ATA.template selfadjointView<Eigen::Upper> ();
+
+  // Approximate as a linear least squares problem
+  for (; source_it.isValid () && target_it.isValid (); ++source_it, ++target_it)
+  {
+    Vector3 s (source_it->x, source_it->y, source_it->z);
+    Vector3 d (target_it->x, target_it->y, target_it->z);
+    Vector3 n (source_it->normal[0] + target_it->normal[0],
+               source_it->normal[1] + target_it->normal[1],
+               source_it->normal[2] + target_it->normal[2]);
+
+    if (!s.array().isFinite().all() ||
+        !d.array().isFinite().all() ||
+        !n.array().isFinite().all())
+    {
+      continue;
+    }
+
+    Vector6 v;
+    v << (s + d).cross (n), n;
+    M.rankUpdate (v);
+   
+    ATb += v * (d - s).dot (n);
+  }
+
+  // Solve A*x = b
+  const auto x = static_cast<Vector6> (M.ldlt ().solve (ATb));
```

Comment:
This looks strange to me. Wouldn't `const Vector6 x = M.ldlt ().solve (ATb)` work?

Reply:
Yeah, that works